modules/promotion/src/Element/CouponRedemptionForm.php

```diff
@@ -224,12 +271,95 @@ public static function validateForm(array &$element, FormStateInterface $form_st
       $form_state->setErrorByName($code_path, t('Coupon is invalid'));
       return;
     }
-    if (!$coupon->getPromotion()->applies($order)) {
-      $form_state->setErrorByName($code_path, t('Coupon is invalid'));
-      return;
-    }
```

Comment:
Because I thought Coupon available triggered that. but I guess not.

Coupon::available checks: status, promotion available
Promotion::available checks: bundle, stores, start/end time, usage limit.

So I guess we still need that.